src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/EngineGeneralTest.java

```diff
@@ -640,6 +646,73 @@ public void batchInsertAutocommitTest() throws Exception {
         assertEquals("COL5  ok?", "OLA", query.get(1).get("COL5").toString());
     }
 
+    /**
+     * Tests that on a rollback situation, the prepared statement batches are cleared.
+     *
+     * The steps performed on this test are:
+     * <ol>
+     *     <li>Add batch to transaction and purposely fail to flush</li>
+     *     <li>Ensure the existence of the Exception and rollback transaction</li>
+     *     <li>Flush again successfully an ensure that the DB table doesn't have any rows</li>
+     * </ol>
+     *
+     * This is a regression test.
+     *
+     * @throws DatabaseEngineException If there is a problem on {@link DatabaseEngine} operations.
+     * @since 2.1.12
+     */
+    @Test
+    public void batchInsertRollback() throws DatabaseEngineException {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        final DbEntity entity = dbEntity()
+                .name("TEST")
+                .addColumn("COL1", INT)
+                .build();
+
+        new MockUp<AbstractDatabaseEngine>() {
+            @Mock
+            public synchronized void flush(final Invocation invocation) throws DatabaseEngineException {
+                if (latch.getCount() == 1) {
+                    throw new DatabaseEngineException("");
+                }
+                invocation.proceed();
+            }
+        };
+
+        DatabaseEngineException expectedException = null;
+
+        engine.addEntity(entity);
+        engine.beginTransaction();
+
+        try {
+            final EntityEntry entry = entry().set("COL1", 1).build();
+
+            engine.addBatch("TEST", entry);
+            engine.flush();
+        } catch (final DatabaseEngineException e) {
+            expectedException = e;
+        } finally {
+            if (engine.isTransactionActive()) {
+                engine.rollback();
+            }
+        }
+
+        // Ensure we had an exception and therefore we didn't insert anything on the DB and that we cleared the batches.
+        assertNotNull("DB returned exception when flushing", expectedException);
+
+        latch.countDown();
+        engine.beginTransaction();
+        engine.flush();
+        engine.commit();
+
+        final List<Map<String, ResultColumn>> query = engine.query(select(all()).from(table("TEST")).orderby(column("COL1").asc()));
+
+        // Previously, we rolled back the transaction; now we are trying the flush an empty transaction.
+        // Therefore, we shouldn't have any rows on the table.
+        assertEquals("There is no rows on table TEST", query.size(), 0);
```

Comment:
"There _are_ no rows..."
and in assertEquals, the "expected" parameter is before the "actual"